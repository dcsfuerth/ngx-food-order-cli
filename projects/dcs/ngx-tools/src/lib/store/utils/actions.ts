export interface IAsyncActionNames {
  base: string;
  start: string;
  success: string;
  error: string;
  complete: string;
  reset: string;
}

/**
 * Generates action names via naming convention .
 *
 * With the given baseName 'PRODUCT' it returns the object
 * ```
 * {
 *   base:  'PRODUCT',
 *   start: 'PRODUCT_START',
 *   next:  'PRODUCT_NEXT',
 *   error: 'PRODUCT_ERROR'
 * }
 * ```
 *
 * @export
 * @param baseName
 */
export function generateAsyncActionNames(baseName: string): IAsyncActionNames {
  return {
    base: baseName,
    start: baseName + ' START',
    success: baseName + ' SUCCESS',
    error: baseName + ' ERROR',
    complete: baseName + ' COMPLETE',
    reset: baseName + ' RESET',
  };
}
