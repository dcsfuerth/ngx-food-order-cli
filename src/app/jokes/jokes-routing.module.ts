import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { JokesPageComponent } from './jokes-page/jokes-page.component';

const routes: Routes = [{ path: '', component: JokesPageComponent }];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class JokesRoutingModule {}
