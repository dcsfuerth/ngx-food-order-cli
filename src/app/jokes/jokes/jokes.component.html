<div class="row">
  <div class="col">
    <h2>Randon Chuck Norris Jokes</h2>
    <ul class="list-group">
      <li *ngFor="let joke of jokes;" class="list-group-item">
        <p [innerHTML]="joke.joke"></p>
        <small>{{ joke.categories }}</small>
      </li>
    </ul>

    <dcs-loading-indicator [loading]="loading"></dcs-loading-indicator>
  </div>
</div>
