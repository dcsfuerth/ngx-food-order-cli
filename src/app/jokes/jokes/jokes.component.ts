import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { Joke } from '../../reducers/jokes/joke.class';

@Component({
  selector: 'dcs-jokes',
  templateUrl: './jokes.component.html',
  styleUrls: ['./jokes.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class JokesComponent {
  @Input() jokes: Joke[];
  @Input() loading: boolean;
}
