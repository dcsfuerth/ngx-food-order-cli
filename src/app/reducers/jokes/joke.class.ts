import { ViewModel } from '@dcs/ngx-tools';

export interface IJoke {
  id: string;
  joke: string;
  categories: string[];
}

export class Joke extends ViewModel<IJoke> {
  protected static readonly defaults: IJoke = {
    id: '',
    joke: '',
    categories: ['nerdy'],
  };

  public id: string;
  public joke: string;
  public categories: string[];

  constructor(props: Partial<IJoke> = {}) {
    props = { ...Joke.defaults, ...props };
    super(props);
  }
}
