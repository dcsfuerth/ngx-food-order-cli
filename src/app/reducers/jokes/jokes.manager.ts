import { State } from '..';
import { schema } from 'normalizr';
import { IJoke, Joke } from './joke.class';
import {
  INormalizedCollectionState,
  generateNormalizedCollectionState,
  readOnlyCollectionManagerFactory,
} from '@dcs/ngx-tools';

export interface IJokesState extends INormalizedCollectionState {
  entities: { jokes: { [key: string]: IJoke } };
}

const initialState = generateNormalizedCollectionState<IJokesState>({ jokes: {} });

const jokeSchema = new schema.Entity(
  'jokes',
  {},
  {
    idAttribute(entity) {
      return String(entity.id);
    },
  }
);

const jokesSchema = new schema.Array(jokeSchema);

export const jokesStateSelector = (state: State): IJokesState => state.jokes;

export const jokesManager = readOnlyCollectionManagerFactory(
  'Jokes',
  initialState,
  jokesStateSelector,
  jokesSchema,
  Joke
);
