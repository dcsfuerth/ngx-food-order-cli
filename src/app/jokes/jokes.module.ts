import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { JokesPageComponent } from './jokes-page/jokes-page.component';
import { JokesRoutingModule } from './jokes-routing.module';
import { JokesComponent } from './jokes/jokes.component';
import { ComponentsModule } from '../components/components.module';

@NgModule({
  imports: [CommonModule, JokesRoutingModule, ComponentsModule],
  declarations: [JokesPageComponent, JokesComponent],
})
export class JokesModule {}
