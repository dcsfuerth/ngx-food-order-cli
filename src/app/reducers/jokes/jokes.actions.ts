import { ApiGetRequest } from '@dcs/ngx-tools';
import { IJoke } from './joke.class';
import { jokesManager } from './jokes.manager';

export interface IJokesResponse {
  type: string;
  value: IJoke[];
}

export class FetchJokes extends ApiGetRequest {
  constructor(count: number = 5) {
    super(
      `http://api.icndb.com/jokes/random/${count}`,
      jokesManager.actions.fetch.base,
      jokesManager.schema,
      undefined,
      // the real data is embedded, unpack before giving it
      // to the normalized tool chain
      (data: IJokesResponse) => data.value
    );
  }
}
