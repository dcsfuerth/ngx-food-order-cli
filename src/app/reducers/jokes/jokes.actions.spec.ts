import { FetchJokes } from './jokes.actions';

describe('Jokes Actions', () => {
  describe('FetchJokes', () => {
    let subject: FetchJokes;

    beforeAll(() => {
      subject = new FetchJokes();
    });

    it('generates the correct API action format', () => {
      expect(subject).toMatchSnapshot();
    });
  });
});
