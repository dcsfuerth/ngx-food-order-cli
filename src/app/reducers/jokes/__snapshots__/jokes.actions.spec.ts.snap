// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jokes Actions FetchJokes generates the correct API action format 1`] = `
FetchJokes {
  "payload": Object {
    "cancel": undefined,
    "handlers": "[Jokes] Fetch",
    "normalizrSchema": ArraySchema {
      "schema": EntitySchema {
        "_getId": [Function],
        "_idAttribute": [Function],
        "_key": "jokes",
        "_mergeStrategy": [Function],
        "_processStrategy": [Function],
        "schema": Object {},
      },
    },
    "rawDataProcessor": [Function],
    "request": Object {
      "method": "GET",
      "url": "http://api.icndb.com/jokes/random/5",
    },
  },
  "type": "[ApiRequest] Execute API Request",
}
`;
